### PR TITLE
Add iterateAll method to Page, update example and snippet

### LINKED ISSUE #330: Provide a convenient way to get all results from a com.google.gcloud.Page
In cases when paging between different requests is not necessary (all results are consumed in the same request/thread) a common usage-pattern would be as such:
```java
Page<V> page = .... get first page
while (page != null) {
  for (V v : page.values()) {
    ...do something...
  }
   page = page.nextPage();
}
```

or
```java
Page<V> page = .... get first page
do {
  for (V v : page.values()) {
    ...do something...
  }
  page = page.nextPage();
} while (page != null);
```

To avoid this boilerplate code It would be nice to add to the `Page` class a method to return
an `Iterator` for all the results (and let that Iterator do the paging).

Some options:

(1) add a static method such as Page.iterateAll(page)
(2) an instance method such as `page.allValues()` or `page.iterateAll()` or `page.fetchAll()`

Any preference or other suggestions?

### PR DESCRIPTION
**Pull request by mziccard**
This PR fixes #330.
- Add an `iterateAll` method to `Page` interface and implements the iterator in `PageImpl`.
- Add unit test
- Update snippet and storage example

### CODE REVIEW

File: gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java [hunk 1]
```diff
@@ -35,6 +37,50 @@
     Page<T> nextPage();
   }
 
+  static class PageIterator<T> implements java.util.Iterator<T> {
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java]: Could be simplified by extending  [AbstractIterator](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/AbstractIterator.html)

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/RemoteGcsHelperTest.java [hunk 2]
```diff
@@ -90,6 +91,11 @@ public String nextPageCursor() {
     public Iterable<BlobInfo> values() {
       return BLOB_LIST;
     }
+
+    @Override
+    public Iterator<BlobInfo> iterateAll() {
+      return null;
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/RemoteGcsHelperTest.java]: BLOB_LIST.iterator() (Though not being used right now)?

**Comment by mziccard** [comment]
Fixed in place.

File: gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java [hunk 3]
```diff
@@ -35,6 +38,33 @@
...
+
+    @Override
+    protected T computeNext() {
+      if (currentPageIterator.hasNext()) {
+        return currentPageIterator.next();
+      }
+      Page<T> nextPage = currentPage.nextPage();
+      if (nextPage != null) {
+        currentPage = nextPage;
+        currentPageIterator = currentPage.values().iterator();
```
Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java]: "Once the implementation either invokes `endOfData` or throws an exception, `computeNext` is guaranteed to never be called again."

So,  I think it should be safe to assign `currentPage.nextPage()` directly to currentPage.

Also, we should change the `if` to a `while` to support the case of an "empty" nextPage that is not last (which theoretically may happen). Maybe something like:
```java
while (!currentPageIterator.hasNext()) {
  currentPage = currentPage.nextPage();
  if (currentPage == null) {
    return endOfData();
  }
  currentPageIterator = currentPage.values().iterator();
}
return currentPageIterator.next();
```

Comment (mziccard) [inline gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java]: Do we really need to support such a case? 
The case with a non null page with empty values should already be impossible. Empty pages in a row seem just an error situation to me.

Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java]: Yes, I think it is possible (I know for sure that it is possible in the case of Datastore but this does not use `Page` right now).

Consider this case: listing with a filter.
Service use a cursor to jump to the start place and then iterate over the result to find values that match the filter criteria. At some point the service might give up before any row matched the criteria returning an empty result but with a cursor to allow continuing the listing from the point it stopped.

I admit that this should not be common but as possibility exists any reason for not supporting it?

Comment (aozarov) [inline gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java]: BTW, another use case (probably even less likely but still possible) is when we deal with a versioned backend storage where one row may have too many deleted versions to be scanned in one request.

**Comment by mziccard** [comment]
Sorry to insist but I am just trying to understand.
> At some point the service might give up before any row matched the criteria returning an empty result but with a cursor to allow continuing the listing from the point it stopped.

According to which criteria? Is this documented somewhere?
In the GCS docs I read that the `pageToken` points to the last item in the *returned list* (i.e. not the last seen before giving up):
> The pageToken is an encoded field representing the name and generation of the last object in the returned list.

So in your example what does prevent the next page request to give up again at the same point returning an empty result list?

I am being picky because I don't want to add a loop if I am not sure according to which criteria it terminates.

**Comment by aozarov** [comment]
No problem. The questions are great!

> In the GCS docs I read that the pageToken points to the last item in the returned list (i.e. not the last seen before giving up):
 
Yes, you are right. I also verified that with @BrandonY a while back.
In the case of GCS it will always return at least one value *or* it will timeout.
I think that is a poor choice of the service and would rather have it behave like Datastore
which can make progress even without returning a result (due to time limit per request), and
let the timeout option be done by the client-side.

> So in your example what does prevent the next page request to give up again at the same point 
returning an empty result list?

Nothing stop also the next result to be empty, this is why it is in a loop.

> I am being picky because I don't want to add a loop if I am not sure according to which criteria it terminates.

The only true indication that there are "no-more-results" is a `null` `nextPageCursor `.

**Comment by mziccard** [comment]
> Yes, you are right. I also verified that with @BrandonY a while back.
In the case of GCS it will always return at least one value or it will timeout.

This confused me, indeed. Thanks for the clarifications!
I added the loop.

**Pull request merged**